File: JANUARY 24/30 January/LCS of three strings gfg.cpp

class Solution
{
    public:

        int LCSof3 (string A, string B, string C, int n1, int n2, int n3)
        {
            // your code here
            // vector<vector<vector<int>>> dp(n1+2,vector<vector<int>>(n2+2,vector<int>(n3+2,0)));
            int dp[n1+2][n2+2][n3+2];
            
            for(int i=0; i<=n1; i++)
            {
                for(int j=0; j<=n2; j++)
                {
                    for(int k=0; k<=n3; k++)
                    {
                        if(i==0 || j==0 || k==0)
                            dp[i][j][k] = 0;
                        else if(A[i-1]==B[j-1] && B[j-1]==C[k-1])
                            dp[i][j][k] = 1+dp[i-1][j-1][k-1];
                        else
                            dp[i][j][k] = max({dp[i-1][j][k], dp[i][j-1][k], dp[i][j][k-1]});
                    }
                }
            }
            return dp[n1][n2][n3];
        }
};

File: JUNE 2025/04 June/LCS of three strings gfg.java
#define vi vector<int>
#define vvi vector<vi>
#define vvvi vector<vvi>
class Solution {
public:
    int lcsOf3(string& s1, string& s2, string& s3) {
        int n = s1.length(), m = s2.length(), o = s3.length();
        
        // Initialize a 3D DP table with (n+1) x (m+1) x (o+1) dimensions, filled with 0s
        vvvi dp(n + 1, vvi(m + 1, vi(o + 1, 0)));
        
        // Iterate through all characters of s1, s2, and s3
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                for (int k = 1; k <= o; ++k) {
                    
                    // If all three characters match
                    if (s1[i - 1] == s2[j - 1] && s2[j - 1] == s3[k - 1]) {
                        dp[i][j][k] = dp[i - 1][j - 1][k - 1] + 1;
                    } 
                    // Else take the maximum of the three possible previous states
                    else {
                        dp[i][j][k] = max({dp[i - 1][j][k], dp[i][j - 1][k], dp[i][j][k - 1]});
                    }
                }
            }
        }
        
        // The answer is stored in dp[n][m][o]
        return dp[n][m][o];
    }
};
